### ISSUE TITLE
Re-using RequestOptions causes wrong HTTP method to be used

### ISSUE THREAD

**Issue by niekvanderkooy** [microsoft/playwright-java#1230]
When doing a request using `requestContext.get("endpoint", requestOptions)` (or `post`, `put`, `delete`, etc), the HTTP method is set using `ensureOptions(requestOptions, "METHOD")`

The implementation of `ensureOptions` does not _change_ the HTTP method if it was already set in your `requestOptions` earlier, meaning that if I first do a `get`, and later a `delete` using the same `RequestOptions` object, then the `requestContext.delete("endpoint", requestOptions)` call still performs a `get`.

I would expect a `delete` to be performed in this second request.

I am using playwright `1.31.0`.

**Comment by yury-s** [comment]
This is grey area, what you are saying is that `delete` etc should ignore `options.method` while current implementation prefers the method from the options. I'd say in this case we should throw if the method doesn't match but it's not very useful for the user. If you always specify method in the options, you should use [`fetch`](https://playwright.dev/java/docs/api/class-apirequestcontext#api-request-context-fetch) which is free of such problems. Any reason you can't use `APIRequestContext.fetch`?

**Comment by niekvanderkooy** [comment]
I have the feeling my problem is the other way around. 

I am not trying to specify my method in the options - I would like to define my method by using `APIRequestContext.delete` etc. However, a side-effect seems to be that this `APIRequestContext.delete` call changes my options object to include `method=delete`, so a subsequent call to `APIRequestContext.get` for example actually ignores the fact that I called `.get`, but prefers the options that I have passed (in which the `method` is now set as a side effect of my call to `delete`).

Looking at the code of `APIRequestContextImpl.ensureOptions`, I am wondering if what I am describing was maybe also the intention, but this method just has a subtle bug? The first line of that method ( `RequestOptionsImpl impl = (RequestOptionsImpl)options;` ) might have been intended to clone the `options` object, and have a specific object `impl` to use in that request. However, since we are not actually cloning but only copying a reference to the passed `options` object, we are actually causing a side effect which is causing my problem.

In that case I could still argue that if explicitly requesting `APIRequestContext.delete`, the method specified should have precedence over the method in the options. But at least if we clone the object during this call, it won't affect any subsequent calls using an object we never intended to change.

I hope I'm being clear, otherwise I could write up a proof of concept of what I'm experiencing tomorrow :)
